src/java/org/apache/cassandra/index/sai/disk/hnsw/CassandraOnHeapHnsw.java

```diff
@@ -311,14 +364,68 @@ public SegmentMetadata.ComponentMetadataMap writeData(IndexDescriptor indexDescr
                 if (vectorPostings.shouldAppendDeletedOrdinal())
                     deletedOrdinals.add(vectorPostings.getOrdinal());
             }
+
+            BiMap <Integer, Integer> ordinalMap = HashBiMap.create();
+            boolean canFastFindRows = false;
+            if (deletedOrdinals.size() == 0) {
+                canFastFindRows = true;
+                int minRow = Integer.MAX_VALUE;
+                int maxRow = Integer.MIN_VALUE;
+                for (VectorPostings<T> vectorPostings : postingsMap.values())
+                {
+                    if (vectorPostings.getRowIds().size() != 1)
+                    {
+                        canFastFindRows = false;
+                        ordinalMap.clear();
+                        break;
+                    }
+                    int rowId = vectorPostings.getRowIds().getInt(0);
+                    int ordinal = vectorPostings.getOrdinal();
+                    minRow = Math.min(minRow, rowId);
+                    maxRow = Math.max(maxRow, rowId);
+                    if(rowId != ordinal)
+                    {
+                        if (ordinalMap.containsKey(ordinal) && ordinalMap.get(ordinal) != rowId)
+                        {
+                            canFastFindRows = false;
+                            ordinalMap.clear();
+                            break;
+                        } else {
+                            ordinalMap.put(ordinal, rowId);
+                        }
+                    }
+                }
+
+                if (canFastFindRows && (minRow != 0 || maxRow != postingsMap.values().size() - 1))
+                {
+                    canFastFindRows = false;
+                    ordinalMap.clear();
+                }
+            }
+
+            boolean needToRemapOrdinals = canFastFindRows && !ordinalMap.isEmpty();
+
+            // write vectors
+            long vectorOffset = vectorsOutput.getFilePointer();
+            long vectorPosition = needToRemapOrdinals
+                                  ? new ReorderingRamAwareVectorValues(vectorValues, ordinalMap).write(vectorsOutput.asSequentialWriter())
+                                  : vectorValues.write(vectorsOutput.asSequentialWriter());
+            long vectorLength = vectorPosition - vectorOffset;
+
             // write postings
             long postingsOffset = postingsOutput.getFilePointer();
-            long postingsPosition = new VectorPostingsWriter<T>().writePostings(postingsOutput.asSequentialWriter(), vectorValues, postingsMap, deletedOrdinals);
+
+            long postingsPosition = canFastFindRows
```

Comment:
would it be simpler if we had a remapping Interface with a `int remap(int ordinal)` method, and just create a default one that always returns ordinal -> ordinal, instead of having if statements everywhere?

(not a rhetorical question, it's possible that the answer is "No, it wouldn't be simpler" because of other factors I don't see yet)

Reply:
I decided to keep it consistent with HnswGraphWriter where it is simpler to pass the BiMap (as direct and inverse mappings are needed)